### DISCUSSION TITLE
Anbindung JK BMS an WT32-ETH01 über input only GPIO möglich?

### DISCUSSION THREAD

**Discussion by peff74** [hoylabs/OpenDTU-OnBattery#353] [General]
Hi,
Ich nutze ja den WT32-ETH01, weil ich eine LAN Anbindung bevorzuge.
Nun frage ich mich, ob ich die wohl bald zur Verfügung stehende JKBMS Funktion nutzen kann.
Denn leider habe ich nur noch GPIO12 frei, ansonsten nur noch read only GPIOs 35/36/39.
Soweit ich verstanden habe, könnte man die für RX nutzen, die Frage ist jetzt, bindet Ihr diese ein?
Werde diese GPIOs auswählbar sein?
Die Frage geht wohl primär an schlimmchen ;-)

Gruß
Peff

**Comment by schlimmchen** [comment]
Ich freu mich, dass du dich für dieses Feature interessierst :blush:

Weil das JK BMS nicht spontan Werte auf seine UART/RS485 Schnittstelle wirft, sondern dazu aufgefordert werden muss, brauchst du genau einen Pin als TX, der auch Output kann. Wenn du noch einen hast, reicht das. Der TX Pin kann ein input-only Pin sein, kein Problem.

Welche Pins verwendet werden, legst du in der `pin_mapping.json` fest. Die sollte dir geläufig sein.

Welches Protokoll wirst du denn einsetzen wollen? Hast du den RS485-Adapter fürs JK BMS? Schließt du an den ESP32 einen UART<->RS485 Adapter an? Oder planst du RX/TX/GND des JK BMS direkt mit dem ESP32 zu verbinden? Hast du in diesem Fall geprüft, welche UART-TTL-Spannung aus dem JK BMS kommt?

Ich hab den RS485 Adapter vom Hersteller und der ist direkt angeschlossen an den auf dem OpenDTU-Fusion v2 integrierten RS485 Transceiver.

**Comment by peff74** [comment]
Moin,
das hört sich doch  gut an, war mir nur nicht sicher, ob man dann die read only Pins in der pin_mapping.json auswählen kann.
Ich müsste dann wohl nur einen PullDown mit an den Pin bringen, laut Doku des ESPs haben die read only keine.

Mit der Anbindung des JK BMS habe ich mich noch nicht tiefer auseinandergesetzt, da Du genau in dem Moment mit Deiner Implementation kamst ;-)

Was ich bisher gesehen hatte, ist die UART vom JK BMS 3.3 Volt.
u. a. hier
https://github.com/Uksa007/esphome-jk-bms-can
![2023-08-03 07_45_13-Window](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/91614910/5f3163b0-0add-4b0c-b0ce-5c6c645fb1a0)

Wobei zu überlegen wäre, ob man hier nicht auch einfach einen ADUM1201ARZ dazwischen flanscht.
Um das ganze zu trennen, dann bräuchte man sich um die Pegel auch nicht mehr zu kümmern.

**Comment by madmartin** [comment]
Hallo @peff74 ,
ich finde die Bezeichnung "Read Only GPIO" problematisch, in den Datenblättern ist die Bezeichnung "Input only GPIO".
Zur Datei "Pinmapping.json" - die kannst Du für Dich kopieren und nach Deinen Wünschen mit einem Texteditor bearbeiten. Wichtig ist die Syntax von JSON richtig hinzukriegen, das ist nicht so einfach.
Dann wird die Datei hochgeladen, und anschliessend musst Du Deine erstelltes Device-Profil im Device-Manager auswählen. 
Welche GPIO Nummern Du da einträgst ist sozusagen beliebig. Für eine funktionierende serielle Schnittstelle brauchst Du natürlich einen Input-Pin für RX und einen Output-Pin für TX. Ob die Kombination TX=12 und RX=12 funktioniert musst du einfach ausprobieren.


Was den Ausschnitt über CAN-Bus betrifft - Deine Schlussfolgerung mit dem ADMUM1201 ist falsch. Wenn CAN-Bus Kommunikation gefragt ist muß ein CAN-Bus Transceiver benutzt werden, z.B. TJA1050. Das sind andere Pegel als 3,3V TTL. 
Für CAN-Bus gibts keine speziellen Pins, es wird auch nur ein Input-Pin für RX und ein Output-Pin für TX benötigt.

**Reply by peff74** [reply]
Hi @madmartin,

O.K. input only GPIOs.
Mit der »Pinmapping.json« kann ich umgehen. habe ich ja auch für meinen WT32-ETH01 angepasst, da die Verkabung auf meiner PCB anders ist.
Aktuelle habe ich auf dem WT32-ETH01 noch einen „normalen“ PIN frei GPIO12, den würde ich dann für TX nehmen.
RX müsste dann einer der  »Input only GPIO« ( 35/36/39) werden.

Thema CAN-BUS - also wenn ich mich nicht sehr verlesen habe, steht da RS485 und nicht CAN-BUS.

**Comment by schlimmchen** [comment]
> Aktuelle habe ich auf dem WT32-ETH01 noch einen „normalen“ PIN frei GPIO12, den würde ich dann für TX nehmen.
RX müsste dann einer der »Input only GPIO« ( 35/36/39) werden.

:+1:

> Thema CAN-BUS - also wenn ich mich nicht sehr verlesen habe, steht da RS485 und nicht CAN-BUS.

Ich glaube @madmartin ist so verwirrt, wie ich es bis gerade noch war. Das Projekt, das @peff74 verlinkt hat und vom den der Screenshot der README.md stammt (und von dem ein Teil meines Codes für OpenDTU-OnBattery inspiriert ist), ist eine ESP32-basierte Bridge zwischen JK BMS und einem Inverter, der CAN spricht (und noch mehr, z.B. HomeAssistent Integration).

Wir reden hier über die elektrische Anbindung des JK BMS an den ESP32, auf dem OpenDTU-OnBattery läuft. CAN spielt hier keine Rolle.

Im verlinkten Projekt ist sogar eine ASCII art Skizze, die ich hier frech kopiere und anpasse:

```
Version 1 (only over short distances, less robust):
          RS485-TTL
┌────────┐         ┌───────┐
│        │<TX---RX>│       │
│ JK-BMS │<RX---TX>│ ESP32 │
│        │<--GND-->│       │
│        │   3V3-->│       │
└────────┘         └───────┘


Version 2 (also working for longer distances, more robust):
Optional: RS485 between JK-BMS GPS port and ESP32, uses JK RS485 adaptor and RS485 to TTL 3V3 adaptor.

          RS485-TTL              RS485           RS485-TTL
┌────────┐         ┌───────────┐       ┌────────┐         ┌───────┐
│        │<--TX--->│Y    JK   Y│<A---A>│        │<TX---RX>│       │
│ JK-BMS │<--RX--->│W  RS485  W│<B---B>│ RS485  │<RX---TX>│ ESP32 │
│        │<--GND-->│B Adaptor B│<-GND->│ to 3V3 │<--GND-->│       │
│        │<-Bat V->│R          │       │        │<--3V3-->│       │
└────────┘         └───────────┘       └────────┘         └───────┘

```

Bei dieser Gelegenheit teile ich das Setup für die Verwendung an einem OpenDTU Fusion v2 Board (sollte nicht überraschen):

```
Version 3:
          RS485-TTL               RS485
                                    ┌──────────────────────────────┐
┌────────┐         ┌───────────┐    │    ┌───────────┐   ┌───────┐ │
│        │<--TX--->│Y    JK   Y│<A--+--A>│A      RXEN│<->│15     │ │
│ JK-BMS │<--RX--->│W  RS485  W│<B--+--B>│B        RX│<->│16 ESP │ │
│        │<--GND-->│B Adaptor  │    │    │       TXEN│<->│46  32 │ │
│        │<-Bat V->│R         B│<-->│GND │ISL3178E EN│<->│45     │ │
└────────┘         └───────────┘    │    └───────────┘   └───────┘ │
                                    │       OpenDTU Fusion v2      │
                                    └──────────────────────────────┘

```

Version 1 und 2 brauchen dann Einstellung `Jikong (JK) BMS per UART and der MCU`, Version 3 wäre `Jikong (JK) BMS per RS485 Transceiver an der MCU`.

**Reply by peff74** [reply]
Ich würde für mich die Version1 anstreben. Kabellänge 20cm. 
Ach und dann könnte natürlich auch der ADMUM1201 mit zwischen, wenn man es Galvanisch trennen möchte, right?

**Reply by schlimmchen** [reply]
Naja, der nimmt ja zweimal VDD, einmal das vom einen Level und einmal das vom anderen Level. Das BMS speist scheinbar die Batteriespannung aus seinem Port, daher darfst du die auf keinen Fall an den ADMIM1201 anschließen, sondern nimmst zweimal 3V3 vom ESP32. Und dann ist der Sinn, diesen Chip einzusetzen, dahin, oder?

**Reply by peff74** [reply]
Mist...
wie geschrieben noch nicht weiter mit auseinandergesetzt...
![2023-08-03 12_16_16-Uksa007_esphome-jk-bms-can_ ESPHome component to monitor a Jikong Battery Manage](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/91614910/073da2fc-cb3b-438b-a56f-91cd64a3873d)

Ob das dann noch Sinn ergibt ... kann ich mangels technischen Know-how nicht sicher beantworten.
